### PR TITLE
MIGENG-130 VMWorkloadInventoryModel

### PR DESCRIPTION
**Pull request by m-brophy**
There are comments (which could be removed if it's felt they're inappropriate) showing the intended usage of fields.
Also I've provided setX and addX methods for fields which will be populated via traversal of 0-many input payload objects. they can be populated all in one go using the setX methods or incrementally as we traverse using addX methods, depending on how we want to do it. We can delete whichever population methods we don't need later on if necessary.

### CODE REVIEW

File: src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    private String osProductName;
+    //hardware/guest_os_full_name
+    private String guestOSFullName;
+    //vms/has_rdm_disk
+    private boolean hasRdmDisk;
+    //count of nics object within the vms/hardware
+    private Integer nicsCount;
+    //hardware/disks/filename
+    private Collection<String> vmDiskFilenames;
+    private Collection<Collection<Map<String, String>>> systemServices;
```
Comment (jonathanvila) [inline src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java]: This is the structure : 
],
            "system_services": [
              {
                "id": 9669,
                "name": "rhnsd",
                "typename": "linux_initprocess"
              },
So unless it's dynamic I would suggest having a Class representing the fields

File: src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    private Collection<Collection<Map<String,String>>> files;
```
Comment (jonathanvila) [inline src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java]: In my example payload I dont have detail on **files** , but I assume we could have the same as with services

File: src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    //vms/operating_system/product_name
+    private String osProductName;
+    //hardware/guest_os_full_name
+    private String guestOSFullName;
+    //vms/has_rdm_disk
+    private boolean hasRdmDisk;
+    //count of nics object within the vms/hardware
+    private Integer nicsCount;
+    //hardware/disks/filename
+    private Collection<String> vmDiskFilenames;
```
Comment (jonathanvila) [inline src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java:35]: If we are gonna have the list of the names of the disks, we could also have their size, and not only the sum on **diskSpace**
What would be the use of having only the list of harddisk names ?

File: src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+import java.util.Collection;
+import java.util.Map;
+
+
+public class VMWorkloadInventoryModel
+{
+    //common/name
+    private String provider;
+    //common/ems_clusters/v_parent_datacenter
+    private String datacenter;
```
Comment (jonathanvila) [inline src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java]: Which one of the ems_clusters would be referred ?

File: src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    //common/ems_clusters/name
+    private String cluster;
```
Comment (jonathanvila) [inline src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java]: which one of the ems_clusters element would be selected to get the name ?

File: src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+    //vms/name
+    private String vmName;
```
Comment (jonathanvila) [inline src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java]: which one of the vms element would be selected to get the name ?

**Comment by jonathanvila** [comment]
I assume the structure is VM centric, therefore we would have 1 object of that class per VM in the structure of the payload.
Am I right ?

File: src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java (hunk 1 above)
Comment (m-brophy) [reply src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java]: I'll change this to be 'systemServicesNames',  a collection of Strings containing all the different 'name' values, which is the only field actually required by the rules.

File: src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java (hunk 2 above)
Comment (m-brophy) [reply src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java]: I'll represent each file by a key, value pair consisting of the name and contents elements

File: src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java (hunk 3 above)
Comment (m-brophy) [reply src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java:35]: To keep the data sent to a minimum that's required we've agreed to keep this as it is.

File: src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java (hunk 5 above)
Comment (jonathanvila) [reply src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java]: //vms/ems_ref

File: src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java (hunk 4 above)
Comment (m-brophy) [reply src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java]: the spreadsheet describes how to find the correct ems_cluster

**Comment by m-brophy** [comment]
> I assume the structure is VM centric, therefore we would have 1 object of that class per VM in the structure of the payload.
> Am I right ?

That's correct

File: src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+    //hardware/guest_os_full_name
+    private String guestOSFullName;
+    //vms/has_rdm_disk
+    private boolean hasRdmDisk;
+    //count of nics object within the vms/hardware
+    private Integer nicsCount;
+    //hardware/disks/filename
+    private Collection<String> vmDiskFilenames;
+    private Collection<String> systemServicesNames;
+    private Collection<Map<String,String>> files;
```
Comment (mrizzi) [inline src/main/java/org/jboss/xavier/analytics/pojo/input/VMWorkloadInventoryModel.java]: I feel like having just a Map (without the Collection) would be enough to accomplish what we decided during the review meeting.
Each file will be an entry with `key = file name` and `content = file content`.
In this way we should be able to collect all the files from the payload in a Map (without the "wrapping" Collection)

```suggestion
    private Map<String,String> files;
```

**Review by mrizzi — COMMENTED** [review]
There's just one further change to be applied and then it will be good.

**Review by mrizzi — APPROVED** [review]
thanks for taking care of creating the bean @m-brophy.
As agreed i'm going to approve & merge this PR even if the build doesn't work due to maven.repository.redhat.com outage

**Pull request merged**